src/backtracking/knight_tour.rs

```diff
@@ -0,0 +1,206 @@
+//!
+//! The Knight's Tour is a classic chess problem where the objective is to move a knight to every square on a chessboard exactly once.
+
+/// Represents the KnightTour struct which implements the Knight's Tour problem.
+pub struct KnightTour {
+    board_size: usize,
+    board: Vec<Vec<i32>>,
+    moves: [(i32, i32); 8],
+}
+
+impl KnightTour {
+    /// Constructs a new KnightTour instance with the given board size.
+    ///
+    /// # Arguments
+    ///
+    /// * `size` - The size of the chessboard.
+    ///
+    /// # Returns
+    ///
+    /// A new KnightTour instance.
+    pub fn new(size: usize) -> Self {
+        let board = vec![vec![0; size]; size];
+        let moves = [
+            (2, 1),
+            (1, 2),
+            (-1, 2),
+            (-2, 1),
+            (-2, -1),
+            (-1, -2),
+            (1, -2),
+            (2, -1),
+        ];
+        KnightTour {
+            board_size: size,
+            board,
+            moves,
+        }
+    }
+
+    /// Checks if the given position is safe to move to.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The x-coordinate of the position.
+    /// * `y` - The y-coordinate of the position.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether the position is safe to move to.
+    fn is_safe(&self, x: i32, y: i32) -> bool {
+        x >= 0
+            && y >= 0
+            && x < self.board_size as i32
+            && y < self.board_size as i32
+            && self.board[x as usize][y as usize] == 0
+    }
+
+    /// Recursively solves the Knight's Tour problem.
+    ///
+    /// # Arguments
+    ///
+    /// * `x` - The current x-coordinate of the knight.
+    /// * `y` - The current y-coordinate of the knight.
+    /// * `move_count` - The current move count.
+    ///
+    /// # Returns
+    ///
+    /// A boolean indicating whether a solution was found.
+    fn solve_tour(&mut self, x: i32, y: i32, move_count: i32) -> bool {
+        if move_count == self.board_size as i32 * self.board_size as i32 {
+            return true;
+        }
+
+        for i in 0..8 {
+            let next_x = x + self.moves[i].0;
+            let next_y = y + self.moves[i].1;
+
+            if self.is_safe(next_x, next_y) {
+                self.board[next_x as usize][next_y as usize] = move_count + 1;
+
+                if self.solve_tour(next_x, next_y, move_count + 1) {
+                    return true;
+                }
+                // Backtrack
+                self.board[next_x as usize][next_y as usize] = 0;
+            }
+        }
+
+        false
+    }
+
+    /// Finds the Knight's Tour starting from the specified position.
+    ///
+    /// # Arguments
+    ///
+    /// * `start_x` - The x-coordinate of the starting position.
+    /// * `start_y` - The y-coordinate of the starting position.
+    ///
+    /// # Returns
+    ///
+    /// A tuple containing a boolean indicating whether a tour was found and the tour matrix if found.
```

Comment:
Please explain what is the _tour matrix_.

Reply:
In my cpde, the term _tour matrix_ refers to a matrix representing the solution to the Knight's Tour problem. The Knight's Tour problem is about finding a sequence of moves for a knight on a chessboard such that the knight visits every square exactly once.

In the context of my implementation:

- The `KnightTour` struct represents the implementation of the Knight's Tour problem.
- The `board` field within the `KnightTour` struct is a two-dimensional vector (`Vec<Vec<i32>>`) representing the chessboard. Each element of this matrix holds an integer value representing the order in which the knight visits that particular square. A value of `0` indicates that the square has not been visited yet.
- The `find_tour` method attempts to find a solution to the Knight's Tour problem starting from a specified position on the chessboard. If a solution is found, it returns a tuple containing a Boolean indicating success (`true`) and the tour matrix.
- The tour matrix returned by `find_tour` is essentially the `board` field of the `KnightTour` struct. It represents the sequence of moves made by the knight on the chessboard, with each cell containing the order in which the knight visited that square.

So, the tour matrix is essentially a record of the knight's path across the chessboard, providing a visual representation of the solution to the problem.

Reply:
Do I need to add explanation for _tour matrix_ to the comment.

Reply:
Yes!

Reply:
Done.